File: 1790. Check if One String Swap Can Make Strings Equal/solution.cpp
#include <iostream>
#include <string>

using namespace std;

bool areAlmostEqual(string s1, string s2) {
  if (s1 == s2)
    return 1;

  int count = 0, notMatch = 0, x = 0, y = 0;

  for (int i = 0; i < s1.size(); i++) {
    if (s1[i] == s2[i])
      count++;
    else {
      notMatch++;
      if (x == 0)
        x = i;
      else
        y = i;
    }
  }
  if ((count + notMatch) == s1.size() && s1[x] == s2[y] && s1[y] == s2[x])
    return 1;
  else
    return 0;
}

int main() {
  string str1 = "bank";
  string str2 = "kanb";
  cout << areAlmostEqual(str1, str2);
}

/*
"bank", "kanb" = true
"attack", "defend" = false
"kelb", "kelb" = true
*/

File: 1790. Check if One String Swap Can Make Strings Equal/solution.js
var areAlmostEqual = function (s1, s2) {
	if (s1 === s2) return true;

	let letters = [];

	for (let i = 0; i < s1.length; i++) {
		if (s1[i] != s2[i]) letters.push(i);
		if (letters.length > 2) return false;
	}

	if (s1[letters[0]] === s2[letters[1]] && s1[letters[1]] === s2[letters[0]]) return true;

	return false;

	/* Another option, which will .push() ALL the differences, and then checkes if 0 == 3 and 1 == 2
    let diff = [];    
    
    for (const i in s1)
        if (s1[i] !== s2[i])
            diff.push(s1[i], s2[i]);
    
    return (diff.length < 5) && (diff[0] === diff[3]) && (diff[1] === diff[2]);
    */
};

const str1 = 'bank';
const str2 = 'kanb';
console.log(areAlmostEqual(str1, str2));

/*
"bank", "kanb" = true
"attack", "defend" = false
"kelb", "kelb" = true
*/
